js/util/pipeline-fixed.js

```diff
@@ -0,0 +1,385 @@
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ * A copy of the GNU Lesser General Public License is in the file COPYING.
+ */
+
+/** @ignore */
+var Interest = require('../interest.js').Interest; /** @ignore */
+var Blob = require('./blob.js').Blob; /** @ignore */
+var KeyChain = require('../security/key-chain.js').KeyChain; /** @ignore */
+var NdnCommon = require('./ndn-common.js').NdnCommon;
+var DataFetcher = require('./data-fetcher.js').DataFetcher;
+var LOG = require('../log.js').Log.LOG;
+
+
+/**
+ * Retrieve the segments of solicited data by keeping a fixed-size window of N
+ * of on fly Interests at any given time.
+ *
+ * To handle timeout and nack we use DataFetcher class which upon facing
+ * timeout or nack will try to resolve the corresponded segment by retransmitting
+ * the Interest a few times.
+ *
+ * PipelineFixed assumes that the data is named /<prefix>/<version>/<segment>,
+ * where:
+ * - <prefix> is the specified name prefix,
+ * - <version> is an unknown version that needs to be discovered, and
+ * - <segment> is a segment number. 
+ * Note: The number of segments is unknown and is controlled by the
+ *       `FinalBlockId` field in the first retrieved Data packet. So, the
+ *       very first Data packet MUST contain the `FinalBlockId` field.
+ *
+ * The following logic is implemented in PipelineFixed:
+ *
+ * 1. Express the first Interest to discover the version:
+ *
+ *    >> Interest: /<prefix>?MustBeFresh=true
+ *
+ * 2. Infer the latest version of the Data: <version> = Data.getName().get(-2)
+ *
+ * 3. If the segment number in the retrieved packet == 0, go to step 5, while the
+ *    next expected segment to fetch will be 1.
+ *
+ * 4. Pipeline the Interets starting from segment 0:
+ *
+ *    >> Interest: /<prefix>/<version>/<segment=0>
+ *    >> Interest: /<prefix>/<version>/<segment=1>
+ *    ...
+ *    >> Interest: /<prefix>/<version>/<segment=this.windowSize-1>
+ *
+ * At any given time the number of on the fly Interests should be equal
+ * to this.windowSize. The next expected segment to fetch will be this.windowSize
+ *
+ * 5. Pipeline the Interests starting from the next expected segment (i.e. M).
+ *
+ *    >> Interest: /<prefix>/<version>/<segment=M>
+ *    >> Interest: /<prefix>/<version>/<segment=M+1>
+ *    ...
+ *    >> Interest: /<prefix>/<version>/<segment=M+this.windowSize-1>
+ *
+ * 6. Upon receiving a valid Data back we pipeline an Interest for the
+ *    next expected segment.
+ *
+ *    >> Interest: /<prefix>/<version>/<segment=(N+1))>
+ *
+ * We repeat step 6 until the FinalBlockId >= Data.getName().get(-1).
```

Comment:
This logic is wrong. Packet Format only guarantees equality comparison between FinalBlockId and last name segment. You can't rely on `>=` operator.

Reply:
Is this only about comparing FinalBlockId and the segment name or you are saying in general we cannot use operators like `>=` `<=` for `FinalBlockId`?? (i.e., is `FinalBlockId >= 10`correct?)